util/settings/accounts.go

```diff
@@ -0,0 +1,307 @@
+	accounts, err := mgr.GetAccounts()
+	if err != nil {
+		return nil, err
+	}
+	account, ok := accounts[name]
+	if !ok {
+		return nil, status.Errorf(codes.NotFound, "account '%s' does not exist", name)
+	}
+	return &account, nil
+}
+
+func (mgr *SettingsManager) UpdateAccount(name string, callback func(account *Account) error) error {
+	account, err := mgr.GetAccount(name)
+	if err != nil {
+		return err
+	}
+	err = callback(account)
+	if err != nil {
+		return err
+	}
+	return mgr.saveAccount(name, *account)
+}
+
+// GetAccounts returns list of configured accounts
+func (mgr *SettingsManager) GetAccounts() (map[string]Account, error) {
+	err := mgr.ensureSynced(false)
+	if err != nil {
+		return nil, err
+	}
+	secret, err := mgr.secrets.Secrets(mgr.namespace).Get(common.ArgoCDSecretName)
+	if err != nil {
+		return nil, err
+	}
+	cm, err := mgr.configmaps.ConfigMaps(mgr.namespace).Get(common.ArgoCDConfigMapName)
+	if err != nil {
+		return nil, err
+	}
+	return parseAccounts(secret, cm)
+}
+
+func updateAccountMap(cm *v1.ConfigMap, key string, val string, defVal string) {
+	existingVal := cm.Data[key]
+	if existingVal != val {
+		if val == "" || val == defVal {
+			delete(cm.Data, key)
+		} else {
+			cm.Data[key] = val
+		}
+	}
+}
+
+func updateAccountSecret(secret *v1.Secret, key string, val string, defVal string) {
+	existingVal := string(secret.Data[key])
+	if existingVal != val {
+		if val == "" || val == defVal {
+			delete(secret.Data, key)
+		} else {
+			secret.Data[key] = []byte(val)
+		}
+	}
+}
+
+func saveAccount(secret *v1.Secret, cm *v1.ConfigMap, name string, account Account) error {
+	tokens, err := json.Marshal(account.Tokens)
+	if err != nil {
+		return err
+	}
+	if name == common.ArgoCDAdminUsername {
+		updateAccountSecret(secret, settingAdminPasswordHashKey, account.PasswordHash, "")
+		updateAccountSecret(secret, settingAdminPasswordMtimeKey, account.FormatPasswordMtime(), "")
+		updateAccountSecret(secret, settingAdminTokensKey, string(tokens), "[]")
+		updateAccountMap(cm, settingAdminEnabledKey, strconv.FormatBool(account.Enabled), "true")
+	} else {
+		updateAccountSecret(secret, fmt.Sprintf("%s.%s.%s", accountsKeyPrefix, name, accountPasswordSuffix), account.PasswordHash, "")
+		updateAccountSecret(secret, fmt.Sprintf("%s.%s.%s", accountsKeyPrefix, name, accountPasswordMtimeSuffix), account.FormatPasswordMtime(), "")
+		updateAccountSecret(secret, fmt.Sprintf("%s.%s.%s", accountsKeyPrefix, name, accountTokensSuffix), string(tokens), "[]")
+		updateAccountMap(cm, fmt.Sprintf("%s.%s.%s", accountsKeyPrefix, name, accountEnabledSuffix), strconv.FormatBool(account.Enabled), "true")
+		updateAccountMap(cm, fmt.Sprintf("%s.%s", accountsKeyPrefix, name), account.FormatCapabilities(), "")
+	}
+	return nil
+}
+
+func parseAdminAccount(secret *v1.Secret, cm *v1.ConfigMap) (*Account, error) {
+	adminAccount := &Account{Enabled: true, Capabilities: []AccountCapability{AccountCapabilityApiKey, AccountCapabilityLogin}}
```

Comment:
Hm, do we really want to allow admin account access to API by default? Or is that for backwards compatibility? If so, we should make a deprecation notice, that admin access to API will be disabled from v2 onwards or something similar.

Reply:
Nice catch. Admin account should not be able to generate API keys by default. That functionality does not exist before 1.5 so we are not breaking backward compatibility. Updated the default capabilities.